Core/Core/Grades/ViewModel/GradeFilterViewModel.swift

```diff
@@ -0,0 +1,145 @@
+
+    private func mapGradingPeriod() {
+        guard dependency.isShowGradingPeriod,
+              let gradingPeriodList = dependency.gradingPeriods,
+              !gradingPeriodList.isEmpty else {
+            isShowGradingPeriodsView = false
+            return
+        }
+
+        // Selected Grading Period
+        let defaultGradingPeriod = GradeFilterViewModel.GradePeriod(title: String(localized: "All", bundle: .core))
+        let selectedGradingId = gradeFilterInteractor.selectedGradingId
+        let selectedGrading = gradingPeriodList.first(where: {$0.id ==  selectedGradingId})
+        let currentGradingPeriod = GradeFilterViewModel.GradePeriod(
+            title: selectedGrading?.title,
+            value: selectedGrading
+        )
+        selectedGradingPeriod = selectedGrading == nil ? defaultGradingPeriod : currentGradingPeriod
+
+        gradingPeriods.append(defaultGradingPeriod)
+        let gradePeriods: [GradeFilterViewModel.GradePeriod] = gradingPeriodList.map { .init(title: $0.title, value: $0) }
+        gradingPeriods.append(contentsOf: gradePeriods)
+    }
+
+    private func mapSortByOptions() {
+        let selectedSortById = gradeFilterInteractor.selectedSortById
+        selectedSortByOption = GradeArrangementOptions(rawValue: selectedSortById ?? 0) ?? .groupName
+        sortByOptions = dependency.sortByOptions
+    }
+
+    private func bindSaveButtonStates() {
+        Publishers.CombineLatest(
+            $selectedGradingPeriod,
+            $selectedSortByOption
+        )
+        .map { [initialFilterValue] grading, sortBy in
+            let selectedFilterValue = FilterValue(
+                selectedGradingPeriod: grading,
+                selectedSortBy: sortBy
+            )
+            return initialFilterValue != selectedFilterValue
+        }
+        .assign(to: &$saveButtonIsEnabled)
+    }
+
+    func saveButtonTapped(viewController: WeakViewController) {
+        dependency.selectedGradingPeriodPublisher.accept(selectedGradingPeriod?.value?.id)
+        dependency.selectedSortByPublisher.accept(selectedSortByOption ?? .groupName)
+        let selectedGradingPeriodId = selectedGradingPeriod?.value?.id
+        gradeFilterInteractor.saveGrading(id: selectedGradingPeriodId)
+        gradeFilterInteractor.saveSortByOption(type: selectedSortByOption ?? .groupName)
+        dimiss(viewController: viewController)
+    }
+
+    func dimiss(viewController: WeakViewController) {
+        dependency.router.dismiss(viewController)
+    }
+}
+
+// MARK: - Dependency
+extension GradeFilterViewModel {
+    public struct Dependency {
+        var router: Router
+        var isShowGradingPeriod: Bool
+        var courseName: String?
+        var selectedGradingPeriodPublisher = PassthroughRelay<String?>()
+        var selectedSortByPublisher = CurrentValueRelay<GradeArrangementOptions>(.groupName)
+        var gradingPeriods: [GradingPeriod]?
+        var sortByOptions: [GradeArrangementOptions]
+    }
```

Comment:
Do they need to be `var` instead of `let`?

Reply:
I have changed some of them let and other keep var, to void make large init and some some of those patterers are optional